precompiles/erc20/integration_test.go

```diff
@@ -0,0 +1,787 @@
+
+			// Setup a previous authorization
+			s.setupSendAuthz(grantee.AccAddr, granter.Priv, sdk.NewCoins(tokenCoin))
+
+			// Approve allowance
+			txArgs, approveArgs := s.getTxAndCallArgs(callType, contractAddr)
+			approveArgs.MethodName = auth.ApproveMethod
+			approveArgs.Args = []interface{}{grantee.Addr, common.Big0}
+
+			approveCheck := passCheck.WithExpEvents(auth.EventTypeApproval)
+
+			_, _, err = s.callContractAndCheckLogs(granter.Priv, txArgs, approveArgs, approveCheck)
+			Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+			// Check allowance contains both spend limits
+			authzs, err := s.grpcHandler.GetAuthorizations(grantee.AccAddr.String(), granter.AccAddr.String())
+			Expect(err).ToNot(HaveOccurred(), "failed to get authorizations")
+			Expect(authzs).To(HaveLen(0), "expected different number of authorizations")
+		},
+			Entry(" - direct call", directCall),
+			// NOTE: we are not passing the contract call here because this test case only covers direct calls
+		)
+
+		DescribeTable("it should return an error if approving 0 and no allowance exists", func(callType int) {
+			grantee := s.keyring.GetKey(1)
+			granter := sender
+
+			// Approve allowance
+			txArgs, approveArgs := s.getTxAndCallArgs(callType, contractAddr)
+			approveArgs.MethodName = auth.ApproveMethod
+			approveArgs.Args = []interface{}{grantee.Addr, common.Big0}
+
+			nonPosCheck := failCheck.WithErrContains("cannot approve non-positive values")
+
+			_, _, err = s.callContractAndCheckLogs(granter.Priv, txArgs, approveArgs, nonPosCheck)
+			Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+
+			// Check allowance contains both spend limits
+			authzs, err := s.grpcHandler.GetAuthorizations(grantee.AccAddr.String(), granter.AccAddr.String())
+			Expect(err).ToNot(HaveOccurred(), "failed to get authorizations")
+			Expect(authzs).To(HaveLen(0), "expected different number of authorizations")
+		},
+			Entry(" - direct call", directCall),
+			// NOTE: we are not passing the contract call here because this test case only covers direct calls
+		)
+
+		// TODO: this is passing right now?? Should we allow someone to create an authorization for themselves?
+		DescribeTable("it should return an error if the grantee is the same as the granter", func(callType int) {
+			grantee := sender
+			granter := sender
+			amount := big.NewInt(100)
+
+			// Approve allowance
+			txArgs, approveArgs := s.getTxAndCallArgs(callType, contractAddr)
+			approveArgs.MethodName = auth.ApproveMethod
+			approveArgs.Args = []interface{}{grantee.Addr, amount}
+
+			_, _, err = s.callContractAndCheckLogs(granter.Priv, txArgs, approveArgs, execRevertedCheck)
+			Expect(err).ToNot(HaveOccurred(), "unexpected result calling contract")
+		},
+			Entry(" - direct call", directCall),
+			// NOTE: we are not passing the contract call here because this test case only covers direct calls
+		)
```

Comment:
We should probably stop people from giving themselves approvals to safe some DB space?

Reply:
how does it work on Ethereum?

Reply:
The Solidity ERC20 contract allows it - there is no real sensible use case here though, so I don't know we should have the same behavior..